쥬스 메이커 [STEP 3] paul, rieul

ParkHyeongSeok opened the pull request:
토니! @Monsteel 
폴(@ParkHyeongSeok), 리을(@jwonyLee)입니다.
STEP3 완료해서 PR 요청 드립니다:)

## 작업 내용
- UpdateStockViewController의 UI 배치
 StackView를 최대한 활용한 뒤 필요한 곳에서 Constraints를 주었습니다.
- FruitStore의 재고 표시
 JuiceMakerViewController에서 주입한 FruitStore의 데이터를 활용해 재고를 표시했습니다.
- Stepper Setting
 Storyboard의 각 Stepper를 하나의 IBAction에 바인딩해 sender.tag로 분기해 작업했습니다. 
 또 Stepper의 변경된 값에 FruitStore의 데이터가 변경되도록 sender.value로 분기해 작업했습니다.
 setupStepper()에서는 FruitStore의 데이터를 활용해 Stepper의 초기값을 세팅했습니다.
- Stepper의 값에 따라 UI를 업데이트
 Stepper의 value에 맞게 각 과일의 UI를 업데이트 하도록 updateUI() 구현했습니다.

## 코드를 작성하면서 고민되었던 점
- updateUI() 대신 setupFruitStock()를 재사용해도 Stepper에 따라서 UI가 업데이트 됩니다. 하지만 모든 값이 동시에 업데이트 되기에, 코드를 재사용하는 것이 좋을지, 해당하는 요소만 업데이트 되도록 updateUI()라는 메서드를 새로 만들어야 할지 고민했습니다.

## 해결이 되지 않은 점

## 조언을 얻고 싶은 부분

ParkHyeongSeok committed:
view: UpdateStockViewController의 AutoLayout 적용

ParkHyeongSeok committed:
feat: UpdateStockViewController에 과일 재고 표시

- JuiceMakerViewController에서 UpdateStockViewController로 데이터를 전달하기 위해 프로퍼티 사용

ParkHyeongSeok committed:
feat: UpdateStockViewController의 Stepper 기본값 세팅

ParkHyeongSeok committed:
feat: UpdateStockViewController의 재고 수정 구현

- chageStockSteppers() 구현
- FruitStore의 changeFruitStock() 구현

ParkHyeongSeok committed:
refator: UpdateStockViewController의 updateUI() 구현

- updateUI()를 구현해서 각 과일의 정보를 독립적으로 업데이트
- Number()를 0으로 리팩토링

jwonyLee committed:
fix: JuiceMakerViewController에서 UpdateStockViewController로 fruitStore 전달

jwonyLee committed:
test: 테스트 코드 수정

- 구현체 변경에 따른 테스트 코드 수정

jwonyLee committed:
test: changeStock 초기재고에서 망고의 재고를 3개로 바꿀 수 있다

Monsteel commented on JuiceMaker/JuiceMaker/Controller/UpdateStockViewController.swift:
```diff
@@ -8,13 +8,53 @@
+    }
+    
+    @IBAction func chageStockSteppers(_ sender: UIStepper) {
+        guard let fruit: Fruit = .init(rawValue: sender.tag) else {
```
`Fruit`의 rawvalue 와 tag가 매치된다는 것을 조금 더 명시적으로 표현할 수 있는 방법은 없을까요 👀

Monsteel reviewed (changes requested):
안녕하세요 폴 리을~! ☀️
좋은 아침입니다 🙌

마이너한 코멘트 한개 남겨 보았습니다 👍

그리고 두분과, 고민해보고 싶은 부분이 하나 있는데요 🙋🏻‍♂️
아래의 method에서, fruitStore를 인자로 받는 방법과, 아래처럼 구현한 방법 중 어떤 방법이 효용성이 효용성이 좋을까요🤔  여러분들의 의견이 궁금합니다 🙋🏻‍♂️

```swift
private func updateFruitStock() {
        guard let fruitStore = self.fruitStore else {
            return
        }
        self.strawberryStockLabel.text = "\(fruitStore.fruits[.strawberry, default: Number()])"
        self.bananaStockLabel.text = "\(fruitStore.fruits[.banana, default: Number()])"
        self.pineappleStockLabel.text = "\(fruitStore.fruits[.pineapple, default: Number()])"
        self.kiwiStockLabel.text = "\(fruitStore.fruits[.kiwi, default: Number()])"
        self.mangoStockLabel.text = "\(fruitStore.fruits[.mango, default: Number()])"
    }
```


그리고, UI에 마이너한 이슈가 하나 있는데요, NavigationBar가 짤려서 보이는 문제가 발생합니다 😓
![스크린샷 2022-02-20 오전 11 41 13](https://user-images.githubusercontent.com/52942409/154826252-46c7bd79-8e40-4165-85b4-04d04dbaf127.png)

남겨주신 코드를 작성하면서 고민되었던 점에 대해 제 생각을 적어보자면..! 📝

해당 프로젝트에서는 두분이 적용하신 방법이 충분히 큰 효용이 있다고 생각합니다 👍
만약 변경되어 반영되어야 하는 데이터가 많다면, 차라리 값이 수정될 때 마다`setupFruitStock()` 를 사용하는것이 (상황마다 다르겠지만..)약간의 성능을 주고, 좋은 가독성을 얻을 수 있을것 같아요 🙌

Step3 진행하시느라 수고 많으셨습니다~! 🎉 🎉

jwonyLee committed:
refactor: FruitStepper 구현

- Fruit.rawValue와 UIStepper.tag를 매치하기 위해 Fruit? 를 가지는 FruitStepper 구현

jwonyLee committed:
refactor: JuiceButton 구현

- Juice.rawValue와 UIButton.tag를 매치하기 위해 Juice? 를 가지는 JuiceButton 구현

jwonyLee committed:
view: UpdateStockViewController 네비게이션 바 오토레이아웃 추가

jwonyLee committed:
refactor: 뷰 컨트롤러 의존성 주입

- StoryboardBindable.instantiate 에서 FruitStore를 파라미터로 받을 수 있게 변경
- BaseViewController 구현
  - FruitStore? 를 가지고 있게함

ParkHyeongSeok committed:
refactor: 함수 의존성 주입

- JuiceMakerViewController.updateFruitStock(_ fruitStore: FruitStore)
- UpdateStockViewController의 setupStepper(_ fruitStore: FruitStore), setupFruitStock(_ fruitStore: FruitStore)

ParkHyeongSeok commented:
토니 주말 저녁이네요! ㅎㅎ @Monsteel 
리뷰해주신거 반영해서 작업해 보았습니다:) 
편히 의견주세요!

> fruitStore를 인자로 받는 방법과, 아래처럼 구현한 방법 중 어떤 방법이 효용성이 좋을까요🤔 여러분들의 의견이 궁금합니다 🙋🏻‍♂️

토니의 리뷰를 받아보니 이 부분을 고민했었다는게 기억나네요 ㅠ 현재 코드처럼 작성하면 컴파일 단계에서 의존성을 주입하는 것이고,  파라미터로 받게되면 런타임에 의존성을 주입하게 되는 차이가 있는 것 같습니다! 함수의 재사용성과 유연성을 고려한다면 파라미터로 받는 함수가 더 좋겠다는 생각이 듭니다:)

그래서 해당 함수를 수정해서 런타임 의존성을 가지도록 구현했습니다.
```swift
// JuiceMakerViewController
private func updateFruitStock(_ fruitStore: FruitStore) { }

// UpdateStockViewController
private func setupStepper(_ fruitStore: FruitStore) { }
private func setupFruitStock(_ fruitStore: FruitStore) { }

// StoryboardBindable
static func instantiate(_ fruitStore: FruitStore? = nil) -> Self { }
```

> UI에 마이너한 이슈가 하나 있는데요, NavigationBar가 짤려서 보이는 문제가 발생합니다 😓

rebase 과정에서 관련 코드가 안들어간 것 같습니다 ㅠ 오토레이아웃을 수정했습니다!

> Fruit의 rawvalue 와 tag가 매치된다는 것을 조금 더 명시적으로 표현할 수 있는 방법은 없을까요 👀

말씀하신 코드를 사용한 곳이 한 곳이 더 있어서 다음과 같이 구현해서 명시적으로 표현해봤습니다!
- FruitStepper와 JuiceButton을 제작하고 각각 fruit과 juice에 접근할 수 있는 프로퍼티를 제작했습니다.
- 그래서 다음 함수에 아래와 같이 명시적으로 표현되도록 수정했습니다.
```swift
@IBAction func changeStock(_ sender: FruitStepper) {
        guard let fruit: Fruit = sender.fruit else {
            return
        }
        self.fruitStore?.changeFruitStock(fruit: fruit, to: changedStock)
	// …
}
```
```swift
 @IBAction func orderJuice(_ sender: JuiceButton) {
        guard let juice: Drinkable = sender.juice else {
            return
        }
        
        do {
            guard let juice: Drinkable = try self.juiceMaker?.makeDrink(juice) else {
                return
            }
	     self.showSuccessAlert(for: juice)
	// …
}
```
